src/vset.c

```diff
@@ -0,0 +1,2094 @@
+ *
+ * Return:
+ *   The updated pVector after removal.
+ *   Returns NULL if the last element was removed and the vector was freed. */
+pVector *pvRemoveAt(pVector *pv, uint32_t idx) {
+    assert(pv && pv->len > 0);
+    assert(idx < pv->len);
+    if (pv->len == 1) {
+        /* Last element being removed; delete vector */
+        zfree(pv);
+        return NULL;
+    } else if (idx < pv->len - 1UL)
+        memmove(&pv->data[idx], &pv->data[idx + 1], (pv->len - idx - 1) * PV_ELEM_SIZE);
+    pv->len--;
+    return pvShrinkToFit(pv);
+}
+
+/* Removes the first matching element from the pVector.
+ *
+ * Performs a linear search for the given pointer and removes the first match.
+ * Updates the vector pointer in case a removal was done.
+ *
+ * Arguments:
+ *   pv   - A pointer to the location of the pVector to remove from.
+ *   elem - The element pointer to match and remove.
+ *
+ * Return:
+ *   true in case a removal was made, false otherwise */
+bool pvRemove(pVector **ppv, void *elem) {
+    pVector *pv = *ppv;
+    if (!pv || pv->len == 0) return false;
+
+    for (uint32_t i = 0; i < pv->len; i++) {
+        if (pv->data[i] == elem) {
+            *ppv = pvRemoveAt(pv, i);
+            return true;
+        }
+    }
+    return false;
+}
+
+/* Retrieves the element at the specified index in the pVector.
+ *
+ * Arguments:
+ *   vec - The pVector to retrieve from.
+ *   idx - The index of the element to access.
+ *
+ * Return:
+ *   A pointer to the element at the given index.
+ *   Returns NULL if the vector is NULL or the index is out of bounds. */
+void *pvGet(pVector *vec, uint32_t idx) {
+    assert(vec && idx < vec->len);
+    return vec->data[idx];
+}
+
+/* Frees the memory used by the pVector.
+ *
+ * Arguments:
+ *   pv - The pVector to free.
+ *
+ * Return:
+ *   None. */
+void pvFree(pVector *pv) {
+    if (pv) zfree(pv);
+}
+
+/* Appends an element to the end of the given pVector.
+ *
+ * Parameters:
+ *   pv   - The vector to append to.
+ *   elem - The element to append.
+ *
+ * Returns:
+ *   A (possibly reallocated) pVector with the new element inserted at the end.
+ *
+ * Notes:
+ *   Internally this uses pvInsert() with the current length of the vector,
+ *   effectively appending the element. */
+pVector *pvPush(pVector *pv, void *elem) {
+    return pvInsert(pv, elem, PV_LEN(pv));
+}
+
+/* Removes and optionally returns the last element from the given pVector.
+ *
+ * Parameters:
+ *   pv    - The vector to remove the element from.
+ *   pelem - Optional pointer to store the popped element. Can be NULL.
+ *
+ * Returns:
+ *   A (possibly reallocated) pVector with the last element removed.
+ *
+ * Notes:
+ *   If the vector is empty, the behavior is to remove from index 0 (safe fallback).
+ *   You can pass NULL for `pelem` if you don't need the removed value. */
+pVector *pvPop(pVector *pv, void **pelem) {
+    uint32_t last_idx = PV_LEN(pv) > 0 ? PV_LEN(pv) - 1 : 0;
```

Comment:
Awkward, relying on assertions in called functions.
```suggestion
pVector *pvPop(pVector *pv, void **pelem) {
    assert(PV_LEN(pv) > 0);
    uint32_t last_idx = PV_LEN(pv) - 1;
```